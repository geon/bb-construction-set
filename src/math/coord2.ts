export type Coord2 = {
	readonly x: number;
	readonly y: number;
};

export const origo: Coord2 = { x: 0, y: 0 };

export function add(a: Coord2, b: Coord2): Coord2 {
	return {
		x: a.x + b.x,
		y: a.y + b.y,
	};
}

export function subtract(a: Coord2, b: Coord2): Coord2 {
	return {
		x: a.x - b.x,
		y: a.y - b.y,
	};
}

export function scale(v: Coord2, factor: number): Coord2 {
	return {
		x: v.x * factor,
		y: v.y * factor,
	};
}

export function magnitude(v: Coord2): number {
	return Math.sqrt(v.x * v.x + v.y * v.y);
}

export function distance(a: Coord2, b: Coord2): number {
	return magnitude(subtract(a, b));
}

export function normalize(v: Coord2): Coord2 {
	return scale(v, 1 / magnitude(v));
}

export function interpolate(from: Coord2, to: Coord2, factor: number): Coord2 {
	return add(scale(from, 1 - factor), scale(to, factor));
}

export function equal(a: Coord2, b: Coord2): boolean {
	return a.x === b.x && a.y === b.y;
}

export function floor(coord: Coord2): Coord2 {
	return {
		x: Math.floor(coord.x),
		y: Math.floor(coord.y),
	};
}

export function multiply(a: Coord2, b: Coord2): Coord2 {
	return {
		x: a.x * b.x,
		y: a.y * b.y,
	};
}

export function divide(a: Coord2, b: Coord2): Coord2 {
	return {
		x: a.x / b.x,
		y: a.y / b.y,
	};
}
