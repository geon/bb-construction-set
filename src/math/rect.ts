import { unzipObject } from "../bb/functions";
import { add, Coord2, origo, subtract } from "./coord2";

export type Rect = {
	/** top left */
	readonly pos: Coord2;
	readonly size: Coord2;
};

export function bottomRight(rect: Rect): Coord2 {
	return add(rect.pos, rect.size);
}

export function flexbox(
	sizes: ReadonlyArray<Coord2>,
	direction: "row" | "column",
	gap: number
): ReadonlyArray<Rect> {
	const axis = ({ row: "x", column: "y" } as const)[direction];

	let pos = origo;
	const rects: Array<Rect> = [];
	for (const size of sizes) {
		rects.push({
			pos,
			size,
		});

		const offsetDistance = size[axis] + gap;
		const offsetVector: Coord2 = { x: 0, y: 0, [axis]: offsetDistance };
		pos = add(pos, offsetVector);
	}

	return rects;
}

export function boundingBox(rects: ReadonlyArray<Rect>): Rect | undefined {
	const topLefts = rects.map((rect) => rect.pos);
	const { x: lefts, y: tops } = unzipObject(topLefts);

	const bottomRights = rects.map(bottomRight);
	const { x: rights, y: bottoms } = unzipObject(bottomRights);

	const min = {
		x: Math.min(...lefts),
		y: Math.min(...tops),
	};

	const max = {
		x: Math.max(...rights),
		y: Math.max(...bottoms),
	};

	return { pos: min, size: subtract(max, min) };
}
