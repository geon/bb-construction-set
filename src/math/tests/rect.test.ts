import { describe, expect, test } from "vitest";
import { bottomRight, flexbox } from "../rect";
import { origo } from "../coord2";

test("bottomRight", () => {
	expect(
		bottomRight({
			pos: { x: 123, y: 456 },
			size: { x: 10, y: 1 },
		})
	).toMatchSnapshot();
});

describe("flexbox", () => {
	test("length", () => {
		expect(flexbox([{ x: 1, y: 1 }], "row", 0).length).toStrictEqual(1);
		expect(
			flexbox(
				[
					{ x: 1, y: 1 },
					{ x: 1, y: 1 },
					{ x: 1, y: 1 },
				],
				"row",
				0
			).length
		).toStrictEqual(3);
	});

	test("first pos", () => {
		expect(flexbox([{ x: 1, y: 1 }], "row", 0)[0]?.pos).toStrictEqual(origo);
	});

	test("use size", () => {
		expect(
			flexbox(
				[
					{ x: 1, y: 1 },
					{ x: 2, y: 3 },
				],
				"row",
				0
			).map((x) => x.size)
		).toStrictEqual([
			{ x: 1, y: 1 },
			{ x: 2, y: 3 },
		]);
	});
});
