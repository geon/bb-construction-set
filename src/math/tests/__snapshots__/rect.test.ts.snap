// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bottomRight 1`] = `
{
  "x": 133,
  "y": 457,
}
`;

exports[`boundingBox > random 1`] = `
{
  "pos": {
    "x": -1,
    "y": 1,
  },
  "size": {
    "x": 15,
    "y": 16,
  },
}
`;
